sdk/python/feast/pyspark/launchers.py

```diff
@@ -0,0 +1,304 @@
+        except Exception as err:
+            raise SparkJobFailure(err)
+        return self._output_file_uri
+
+
+class JobLauncher(abc.ABC):
+    """
+    Submits spark jobs to a spark cluster. Currently supports only historical feature retrieval jobs.
+    """
+
+    @abc.abstractmethod
+    def historical_feature_retrieval(
+        self,
+        pyspark_script: str,
+        entity_source_conf: Dict,
+        feature_tables_sources_conf: List[Dict],
+        feature_tables_conf: List[Dict],
+        destination_conf: Dict,
+        job_id: str,
+        **kwargs,
+    ) -> RetrievalJob:
+        """
+        Submits a historical feature retrieval job to a Spark cluster.
+
+        Args:
+            pyspark_script (str): Local file path to the pyspark script for historical feature
+                retrieval.
+            entity_source_conf (List[Dict]): Entity data source configuration.
+            feature_tables_sources_conf (Dict): List of feature tables data sources configurations.
+            feature_tables_conf (List[Dict]): List of feature table specification.
+                The order of the feature table must correspond to that of feature_tables_sources.
+            destination_conf (Dict): Retrieval job output destination.
+            job_id (str): A job id that is unique for each job submission.
+
+        Raises:
+            SparkJobFailure: The spark job submission failed, encountered error
+                during execution, or timeout.
+
+        Returns:
+            str: file uri to the result file.
+        """
+        raise NotImplementedError
+
+
+class StandaloneCluster(JobLauncher):
+    """
+    Submits jobs to a standalone Spark cluster in client mode.
+    """
+
+    def __init__(self, master_url: str, spark_home: str = None):
+        """
+        This launcher executes the spark-submit script in a subprocess. The subprocess
+        will run until the Pyspark driver exits.
+
+        Args:
+            master_url (str):
+                Spark cluster url. Must start with spark://.
+            spark_home (str):
+                Local file path to Spark installation directory. If not provided,
+                the environmental variable `SPARK_HOME` will be used instead.
+        """
+        self.master_url = master_url
+        self.spark_home = spark_home if spark_home else os.getenv("SPARK_HOME")
+
+    @property
+    def spark_submit_script_path(self):
+        return os.path.join(self.spark_home, "bin/spark-submit")
+
+    def historical_feature_retrieval(
+        self,
+        pyspark_script: str,
+        entity_source_conf: Dict,
+        feature_tables_sources_conf: List[Dict],
+        feature_tables_conf: List[Dict],
+        destination_conf: Dict,
+        job_id: str,
+        **kwargs,
+    ) -> RetrievalJob:
+
+        submission_cmd = [
+            self.spark_submit_script_path,
+            "--master",
+            self.master_url,
+            "--name",
+            job_id,
```

Comment:
Why is `job_id` externally configurable here? (as opposed to generating one inside JobLauncher)

Reply:
Feast client will be the one generating the `job_id`, rather than user specified. As to why the job id is generated in Feast Client, rather than JobLauncher:

Prior to launching the job, JobLauncher will be doing some preparatory task, such as uploading Pandas dataframe to GCS / S3 (if user input is a pandas dataframe rather than a uri pointing to the file) , and define the output path based on Feast client configuration. In both cases, we can use job id as part of the GCS / S3 path name, which would be useful for tracking purposes.

That being said, i am fine with moving the job id generation within JobLauncher instead, if there is a good reason as to why it's better to have the job id generation within Job Launcher instead of Feast Client?

Reply:
I see. I don't feel super strongly either way, but you could say that job id generation is a part of JobLauncher internal logic. That way the cloud-specific implementation would be able to use a specific jobid format for convenience. There might be cloud-specific limitations on job id length / charset that Feast Client is not aware of.

It may come in handy given that we don't plan to store jobs in a central database, they will have to be stored as some attribute or tag in dataproc/EMR job metadata itself.

Reply:
Is this `job_id` the internal `job_id` or the external one? They might be different. The latter should be encapsulated, the former is internal to Feast and would be exposed to users. Or are we going to make them one and the same?